backend/tardigrade/object.go

```diff
@@ -0,0 +1,214 @@
+		}
+	}
+
+	bucketName, _ := bucket.Split(path.Join(f.root, relative))
+
+	return &Object{
+		fs: f,
+
+		absolute: norm.NFC.String(bucketName + "/" + object.Key),
+
+		size:     object.System.ContentLength,
+		created:  object.System.Created,
+		modified: modified,
+	}
+}
+
+// String returns a description of the Object
+func (o *Object) String() string {
+	if o == nil {
+		return "<nil>"
+	}
+
+	return o.Remote()
+}
+
+// Remote returns the remote path
+func (o *Object) Remote() string {
+	// This is a special case required by rclone. If the filesystem root is
+	// actually an object, then instead of returning an empty string (which
+	// would be the normal relative path to the file) we must return the
+	// basename of the file.
+	if o.fs.object != nil {
+		_, filename := path.Split(o.fs.root)
+
+		return filename
+	}
+
+	// It is possible that we have an empty root (meaning the filesystem is
+	// rooted at the project level). In this case the relative path is just
+	// the full absolute path to the object (including the bucket name).
+	if o.fs.root == "" {
+		return o.absolute
+	}
+
+	// At this point we know that the filesystem itself is at least a
+	// bucket name (and possibly a prefix path).
+	//
+	//                               . This is necessary to remove the slash.
+	//                               |
+	//                               v
+	return o.absolute[len(o.fs.root)+1:]
+}
+
+// ModTime returns the modification date of the file
+// It should return a best guess if one isn't available
+func (o *Object) ModTime(ctx context.Context) time.Time {
+	return o.modified
+}
+
+// Size returns the size of the file
+func (o *Object) Size() int64 {
+	return o.size
+}
+
+// Fs returns read only access to the Fs that this object is part of
+func (o *Object) Fs() fs.Info {
+	return o.fs
+}
+
+// Hash returns the selected checksum of the file
+// If no checksum is available it returns ""
+func (o *Object) Hash(ctx context.Context, ty hash.Type) (_ string, err error) {
+	fs.Debugf(o, "%s", ty)
+
+	return "", hash.ErrUnsupported
+}
+
+// Storable says whether this object can be stored
+func (o *Object) Storable() bool {
+	return true
+}
+
+// SetModTime sets the metadata on the object to set the modification date
+func (o *Object) SetModTime(ctx context.Context, t time.Time) (err error) {
+	fs.Debugf(o, "touch -d %q sj://%s", t, o.absolute)
+
+	return fs.ErrorCantSetModTime
+}
+
+// Open opens the file for read. Call Close() on the returned io.ReadCloser
+func (o *Object) Open(ctx context.Context, options ...fs.OpenOption) (_ io.ReadCloser, err error) {
+	fs.Debugf(o, "cat sj://%s # %+v", o.absolute, options)
+
+	bucketName, bucketPath := bucket.Split(o.absolute)
+
+	// Convert the semantics of HTTP range headers to an offset and length
+	// that libuplink can use.
+	var (
+		offset int64 = 0
+		length int64 = -1
+	)
+
+	for _, option := range options {
+		switch opt := option.(type) {
+		case *fs.RangeOption:
+			s := opt.Start >= 0
```

Comment:
I think you might find RangeOption.Decode useful here as it returns the offset and length directly

Reply:
Using `RangeOption.Decode` would require knowing the size of the object (and an additional network roundtrip to get it). In theory we could `RangeOption.Decode(0)` under some of the cases, but that implies too much knowledge about the way the decode implementation works.

Maybe we should extend RangeOption with `RequiresSize()` and `SetSize()`? I'm imagining code like:

```
if opt.RequiresSize() {
        object, err := o.fs.project.StatObject(ctx, bucketName, bucketPath)
        if err != nil {
                return nil, err
        }

        opt.SetSize(object.ContentLenth)
}

offset, length := opt.Decode()
```

Reply:
Ah, I missed the subtlety that you were avoiding a stat call...

`RequiresSize()` would work. I was thinking of a function to read the size that you could pass in

```
func (o *RangeOption) DecodeSizeFn(size func() (int64, error)) (offset, limit int64, err error) {
}
```

I think `RequiresSize` is probably neater!

I don't think this is something we need to address before merging!